File: packages/client-swift/Sources/SpriteCoreClient/AnimationGraph.swift
import Foundation

/// Resolved animation table + transition graph for a single mode of a single
/// character. Both sprite and atlas manifests project into this shape so the
/// player stays format-agnostic.
public struct AnimationGraph: Sendable {
    public let defaultState: String
    public let animations: [String: Animation]
    public let transitions: [String: TransitionRef]

    public init(defaultState: String, animations: [String: Animation], transitions: [String: TransitionRef]) {
        self.defaultState = defaultState
        self.animations = animations
        self.transitions = transitions
    }

    /// Resolve a state→state transition against the transitions table using
    /// wildcard pattern matching. Specificity order (most→least specific):
    ///
    ///   `"<from>-><to>"` → `"<from>->*"` → `"*-><to>"` → `"*->*"`
    ///
    /// Returns nil when nothing matches.
    public func resolveTransition(from: String, to: String) -> TransitionRef? {
        let keys = ["\(from)->\(to)", "\(from)->*", "*->\(to)", "*->*"]
        for k in keys {
            if let t = transitions[k] { return t }
        }
        return nil
    }

    /// Extract a single mode's animation graph from a character manifest.
    public static func fromManifest(_ manifest: CharacterManifest, mode: String) throws -> AnimationGraph {
        guard let content = manifest.content[mode] else {
            throw GraphError.modeNotFound(mode: mode, available: Array(manifest.content.keys))
        }
        let defaultState = try resolveDefaultState(stateMap: manifest.stateMap, animations: content.animations)
        return AnimationGraph(
            defaultState: defaultState,
            animations: content.animations,
            transitions: content.transitions ?? [:]
        )
    }
}

public enum GraphError: Error, CustomStringConvertible {
    case modeNotFound(mode: String, available: [String])
    case noAnimations

    public var description: String {
        switch self {
        case .modeNotFound(let mode, let available):
            return "manifest has no content for mode '\(mode)'. Available: \(available)"
        case .noAnimations:
            return "manifest mode has no animations"
        }
    }
}

private func resolveDefaultState(stateMap: [String: String], animations: [String: Animation]) throws -> String {
    for (_, animName) in stateMap {
        if animations[animName] != nil { return animName }
    }
    if let first = animations.keys.first { return first }
    throw GraphError.noAnimations
}

/// The three phases of a phased animation; flat animations use `.loop`.
public enum Phase: String, Sendable {
    case intro
    case loop
    case outro
}

/// A transition target resolved for playback: which animation + phase to
/// play once before entering the target state's own loop.
public struct ResolvedTransition: Sendable {
    public let animation: String
    public let phase: Phase

    /// Parse `"thinking.intro"` → `(thinking, .intro)`. Unqualified → `.loop`.
    public static func parse(_ ref: String) -> ResolvedTransition {
        if let dot = ref.firstIndex(of: ".") {
            let phase = Phase(rawValue: String(ref[ref.index(after: dot)...])) ?? .loop
            return ResolvedTransition(animation: String(ref[..<dot]), phase: phase)
        }
        return ResolvedTransition(animation: ref, phase: .loop)
    }
}

File: packages/client-js/src/animation-graph.ts
import type {
  Animation,
  CharacterManifest,
  TransitionRef,
} from "./schema.js";

/**
 * Resolved animation table + transition graph for a single mode of a single
 * character. Both sprite and atlas manifests project into this shape so the
 * player stays format-agnostic.
 *
 * Build via [fromManifest] to pull a mode's content out of a server-synthesized
 * [CharacterManifest], or construct directly for tests.
 */
export class AnimationGraph {
  constructor(
    readonly defaultState: string,
    readonly animations: Readonly<Record<string, Animation>>,
    readonly transitions: Readonly<Record<string, TransitionRef>>,
  ) {}

  /**
   * Resolve a state→state transition against the transitions table using
   * wildcard pattern matching. Specificity order (most→least specific):
   *
   *   "<from>-><to>" → "<from>->*" → "*-><to>" → "*->*"
   *
   * Returns null when nothing matches; the caller then swaps instantly.
   */
  resolveTransition(from: string, to: string): TransitionRef | null {
    const keys = [`${from}->${to}`, `${from}->*`, `*->${to}`, `*->*`];
    for (const k of keys) {
      const t = this.transitions[k];
      if (t !== undefined) return t;
    }
    return null;
  }

  /**
   * Extract a single mode's animation graph from a character manifest. The
   * default state is taken from [stateMap] — the first key that maps to an
   * animation present in [mode]'s content — or fails if no animation is
   * present.
   */
  static fromManifest(manifest: CharacterManifest, mode: string): AnimationGraph {
    const content = manifest.content[mode];
    if (!content) {
      throw new Error(
        `manifest has no content for mode '${mode}'. Available: [${Object.keys(manifest.content).join(", ")}]`,
      );
    }
    const defaultState = resolveDefaultState(manifest.stateMap, content.animations);
    return new AnimationGraph(defaultState, content.animations, content.transitions ?? {});
  }
}

function resolveDefaultState(
  stateMap: Record<string, string>,
  animations: Record<string, Animation>,
): string {
  for (const [, animName] of Object.entries(stateMap)) {
    if (animName in animations) return animName;
  }
  const first = Object.keys(animations)[0];
  if (first === undefined) {
    throw new Error("manifest mode has no animations");
  }
  return first;
}

/** The three phases of a phased animation; flat animations use `loop`. */
export type Phase = "intro" | "loop" | "outro";

/**
 * A transition target resolved for playback: which animation + phase to play
 * once before entering the target state's own loop. Used by the player when a
 * phase-string `TransitionRef` fires on state change.
 */
export type ResolvedTransition = {
  animation: string;
  phase: Phase;
};

/** Parse `"thinking.intro"` into `{ animation: "thinking", phase: "intro" }`. Unqualified → loop. */
export function resolveTransition(ref: string): ResolvedTransition {
  const dot = ref.indexOf(".");
  if (dot < 0) return { animation: ref, phase: "loop" };
  const phase = ref.slice(dot + 1) as Phase;
  if (phase !== "intro" && phase !== "loop" && phase !== "outro") {
    throw new Error(`unknown phase: ${phase}`);
  }
  return { animation: ref.slice(0, dot), phase };
}

/**
 * Treat a flat animation as the `loop` phase so the player can always look up
 * phases by name without special-casing flat vs phased at every site.
 */
export function effectiveLoop(anim: Animation) {
  return anim.loop ?? anim.sequence;
}
